### DISCUSSION TITLE
How can I embed a youtube video with markdown?

### DISCUSSION THREAD

**Discussion by marcmaurer99** [requarks/wiki#4580] [Help / Questions]
Anybody can help me how I can embed a youtube video with markdown in wiki.js?

Thanks!

Marc

**Comment by Phil-hostwiki** [comment]
Go to Admin Area -> Theme -> Code Injection

Under `Head HTML Injection`
insert the code below.
```
<script type="text/javascript" defer>
  const rxYoutube = /^.*^((?:https?:)?\/\/)?((?:www|m)\.)?((?:youtube\.com|youtu.be))(\/(?:[\w\-]+\?v=|embed\/|v\/|shorts\/)?)([\w\-]+)(\S+)?$/

 const rxScreencast = /^.*^((?:https?:)?\/\/)?(www\.)?(screencast\.com)(\/users)\/([a-z0-9_-]+)\/folders\/([a-z0-9%_-]+)\/media\/([a-z0-9_-]+)(?:\/)?$/im

  window.boot.register('vue', () => {
    window.onload = () => {
      document.querySelectorAll('.contents oembed, .contents a').forEach(elm => {
        const url = elm.hasAttribute('url') ? elm.getAttribute('url') : elm.getAttribute('href')
        let newElmHtml = null
       
        const ytMatch = url.match(rxYoutube)
        const scMatch = url.match(rxScreencast)

 
        if (ytMatch) {
          newElmHtml = `<iframe id="ytplayer" type="text/html" width="640" height="360" src="https://www.youtube-nocookie.com/embed/${ytMatch[5]}" frameborder="0" allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>`
        
      } else if(scMatch) {
        newElmHtml = `<iframe id="scplayer" type="text/html" width="640" height="360" src="${url}/embed" frameborder="0" allowfullscreen></iframe>`

        } else if (url.endsWith('.mp4')) {
          newElmHtml = `<video controls autostart="0" name="media" width="640" height="360"><source src="${url}" type="video/mp4"></video>`
        } else {
         return
         }

        const newElm = document.createElement('div') 
        newElm.classList.add('responsive-embed')
        newElm.insertAdjacentHTML('beforeend', newElmHtml)
        elm.replaceWith(newElm)
      })
    }  
  })
</script>
```

Under `CSS Override`
```
.responsive-embed {
  position: relative;
  padding-bottom: 56.2%;
  height: 0;
  margin: 10px 0;
  overflow: hidden;
}

.responsive-embed iframe {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border-radius: 15px;
}
```

To embed any Youtube video, just paste the link to the video inside the markdown editor. e.g `https://youtube.com/watch?v=qIsgdOVGA04` 

Practical screenshots below:
1.  
![image](https://user-images.githubusercontent.com/82508256/138730159-ecfbd3e1-f804-4687-8f0a-2368ee2be69c.png)

2.  
![image](https://user-images.githubusercontent.com/82508256/138729245-97dc0889-9002-4406-bfce-fd8806b78cd0.png)

3.  
![image](https://user-images.githubusercontent.com/82508256/138730008-a56d389a-0db7-4ca3-8c32-a0d7210e68a2.png)

I hope this helps.

**Reply by marcmaurer99** [reply]
Thanks a lot. Works great with regular videos. Question, this does not work with videos that are set on private (i.e. the link is not publicly accessible). Is there a way to make it work for private videos too?

**Reply by Phil-hostwiki** [reply]
Private videos cannot be embedded. You should set the video to "unlisted".

**Reply by marcmaurer99** [reply]
perfect, thank you so much!

**Reply by peteruithoven** [reply]
Does this also work for the Visual editor > Insert media functionality?

**Reply by Phil-hostwiki** [reply]
> Does this also work for the Visual editor > Insert media functionality? 

It works on the visual editor using the "Link" insert feature.

![image](https://user-images.githubusercontent.com/82508256/169618613-cd75b005-5fda-4ec5-981e-d70ee1c4b689.png)

**Reply by wangjueszu** [reply]
is just use to youtube?

**Reply by asisteclick** [reply]
Hi! I followed the steps detailed here and still the video disappears after exiting edit mode. Any ideas why?

Best,

**Reply by Phil-hostwiki** [reply]
> Hi! I followed the steps detailed here and still the video disappears after exiting edit mode. Any ideas why?

Which of the editors are you using?

**Reply by asisteclick** [reply]
I'm using the Visual Editor, Wiki.js 2.5.289. Also tried disabling html > html > security > allow iframes.

**Reply by Phil-hostwiki** [reply]
Please confirm that you are not using the default “Insert media”, and that you are using the Link insert instead?

**Reply by asisteclick** [reply]
I'm using the Insert Media. I just managed to have it worked out by adding this code instead:

```
<script type="text/javascript" defer>
  const rxYoutube = /^.*(?:(?:youtu\.be\/|v\/|vi\/|u\/\w\/|embed\/)|(?:(?:watch)?\?v(?:i)?=|\&v(?:i)?=))([^#\&\?]*).*/
  const rxVimeo = /^.*(vimeo\.com\/)((channels\/[A-z]+\/)|(groups\/[A-z]+\/videos\/))?([0-9]+)/

  window.boot.register('vue', () => {
    window.onload = () => {
      document.querySelectorAll('.contents oembed, .contents a').forEach(elm => {
        const url = elm.hasAttribute('url') ? elm.getAttribute('url') : elm.getAttribute('href')
        let newElmHtml = null
        
        const ytMatch = url.match(rxYoutube)
        const vmMatch = !ytMatch && url.match(rxVimeo)
        if (ytMatch) {
          newElmHtml = `<iframe id="ytplayer" type="text/html" src="https://www.youtube.com/embed/${ytMatch[1]}" frameborder="0" allowfullscreen allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" ></iframe>`
        } else if (vmMatch) {
          newElmHtml = `<iframe id="vmplayer" type="text/html" width="640" height="360" src="https://player.vimeo.com/video/${vmMatch[5]}" frameborder="0" allowfullscreen></iframe>`
        } else if (url.endsWith('.mp4')) {
          newElmHtml = `<video controls autostart="0" name="media" width="640" height="360"><source src="${url}" type="video/mp4"></video>`
        } else {
          return
        }

        const newElm = document.createElement('div') 
        newElm.classList.add('video-responsive')
        newElm.insertAdjacentHTML('beforeend', newElmHtml)
        elm.replaceWith(newElm)
      })
    }  
  })
</script>
```

**Reply by Phil-hostwiki** [reply]
> I'm using the Insert Media. I just managed to have it worked out by adding this code instead:

The initial code works with link inserts, that is why it didn’t work for you.

**Reply by asisteclick** [reply]
I see! So now we have both ways here for other people to use. Thanks a lot for your help Phil!

**Reply by Phil-hostwiki** [reply]
You are welcome.

**Reply by mittkkoo** [reply]
The script worked for me! Thank you for sharing!

However as a side-effect all my other scripts that I used to insert in the HTML body stopped working. They still appear in the page, but they don't work. I can't figure out why...

**Reply by Philipinho** [reply]
> @mittkkoo: The script worked for me! Thank you for sharing!
> 
> However as a side-effect all my other scripts that I used to insert in the HTML body stopped working. They still appear in the page, but they don't work. I can't figure out why...

Do you mean you have existing Javascript you injected before and they are no longer working?
Are they scripts you can share here so I can figure out how to combine them to work together?
If not, you can share with me on the Wiki.js slack.

I am still @Phil-hostwiki

**Reply by mittkkoo** [reply]
Yes, I my other scripts don't work any more. Actually it is only one. It used to work when inserted in the HTML body. But not any more after I injected the youtube script.

```
<script>
function replaceGermanLocals() {
   var x = document.getElementsByClassName("text-none");
   var i;
   for (i = 0; i < x.length; i++) {
      x[i].innerText = x[i].innerText.replace(/Kommentar veröffentlichen/g, 'OK');
   }
};
window.onload = function() {
  replaceGermanLocals();
};
</script>
```

**Reply by Philipinho** [reply]
Alright, I will have a look.

**Reply by Philipinho** [reply]
@mittkkoo Try this https://pastebin.com/raw/vp87cnkX

**Reply by mittkkoo** [reply]
Works! Now I understand what was going on with window.onload.

**Reply by mittkkoo** [reply]
> @mittkkoo Try this https://pastebin.com/raw/vp87cnkX

Based on a comment by @NGPixel [here](https://github.com/requarks/wiki/discussions/6293#discussioncomment-5423370) I changed `window.onload` to this:
```javascript
  window.boot.register('page-ready', () => {
    replaceGermanLocals();
    handleEmbeddedMedia();    
  })
```

**Reply by samuk10** [reply]
> Go to Admin Area -> Theme -> Code Injection
> 
> Under `Head HTML Injection` insert the code below.
> 
> ```
> <script type="text/javascript" defer>
>   const rxYoutube = /^.*^((?:https?:)?\/\/)?((?:www|m)\.)?((?:youtube\.com|youtu.be))(\/(?:[\w\-]+\?v=|embed\/|v\/|shorts\/)?)([\w\-]+)(\S+)?$/
> 
>  const rxScreencast = /^.*^((?:https?:)?\/\/)?(www\.)?(screencast\.com)(\/users)\/([a-z0-9_-]+)\/folders\/([a-z0-9%_-]+)\/media\/([a-z0-9_-]+)(?:\/)?$/im
> 
>   window.boot.register('vue', () => {
>     window.onload = () => {
>       document.querySelectorAll('.contents oembed, .contents a').forEach(elm => {
>         const url = elm.hasAttribute('url') ? elm.getAttribute('url') : elm.getAttribute('href')
>         let newElmHtml = null
>        
>         const ytMatch = url.match(rxYoutube)
>         const scMatch = url.match(rxScreencast)
> 
>  
>         if (ytMatch) {
>           newElmHtml = `<iframe id="ytplayer" type="text/html" width="640" height="360" src="https://www.youtube-nocookie.com/embed/${ytMatch[5]}" frameborder="0" allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>`
>         
>       } else if(scMatch) {
>         newElmHtml = `<iframe id="scplayer" type="text/html" width="640" height="360" src="${url}/embed" frameborder="0" allowfullscreen></iframe>`
> 
>         } else if (url.endsWith('.mp4')) {
>           newElmHtml = `<video controls autostart="0" name="media" width="640" height="360"><source src="${url}" type="video/mp4"></video>`
>         } else {
>          return
>          }
> 
>         const newElm = document.createElement('div') 
>         newElm.classList.add('responsive-embed')
>         newElm.insertAdjacentHTML('beforeend', newElmHtml)
>         elm.replaceWith(newElm)
>       })
>     }  
>   })
> </script>
> ```
> 
> Under `CSS Override`
> 
> ```
> .responsive-embed {
>   position: relative;
>   padding-bottom: 56.2%;
>   height: 0;
>   margin: 10px 0;
>   overflow: hidden;
> }
> 
> .responsive-embed iframe {
>   position: absolute;
>   top: 0;
>   left: 0;
>   width: 100%;
>   height: 100%;
>   border-radius: 15px;
> }
> ```
> 
> To embed any Youtube video, just paste the link to the video inside the markdown editor. e.g `https://youtube.com/watch?v=qIsgdOVGA04`
> 
> Practical screenshots below: 1. ![image](https://user-images.githubusercontent.com/82508256/138730159-ecfbd3e1-f804-4687-8f0a-2368ee2be69c.png)
> 
> 
> ![image](https://user-images.githubusercontent.com/82508256/138729245-97dc0889-9002-4406-bfce-fd8806b78cd0.png)
> 
> 
> ![image](https://user-images.githubusercontent.com/82508256/138730008-a56d389a-0db7-4ca3-8c32-a0d7210e68a2.png)
> 
> I hope this helps.

Thanks i also managed it to work with google-drive videos!
Aqui está o código completo com a adição do suporte para vídeos do Google Drive:

```
<script type="text/javascript" defer>
  const rxYoutube = /^.*^((?:https?:)?\/\/)?((?:www|m)\.)?((?:youtube\.com|youtu.be))(\/(?:[\w\-]+\?v=|embed\/|v\/|shorts\/)?)([\w\-]+)(\S+)?$/

  const rxScreencast = /^.*^((?:https?:)?\/\/)?(www\.)?(screencast\.com)(\/users)\/([a-z0-9_-]+)\/folders\/([a-z0-9%_-]+)\/media\/([a-z0-9_-]+)(?:\/)?$/im

  const rxDrive = /^.*^((?:https?:)?\/\/)?((?:www|m)\.)?(drive\.google\.com)\/file\/d\/([a-zA-Z0-9_-]+)\/.*$/i

  window.boot.register('vue', () => {
    window.onload = () => {
      document.querySelectorAll('.contents oembed, .contents a').forEach(elm => {
        const url = elm.hasAttribute('url') ? elm.getAttribute('url') : elm.getAttribute('href')
        let newElmHtml = null

        const ytMatch = url.match(rxYoutube)
        const scMatch = url.match(rxScreencast)

        if (ytMatch) {
          newElmHtml = `<iframe id="ytplayer" type="text/html" width="640" height="360" src="https://www.youtube-nocookie.com/embed/${ytMatch[5]}" frameborder="0" allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>`
        } else if (scMatch) {
          newElmHtml = `<iframe id="scplayer" type="text/html" width="640" height="360" src="${url}/embed" frameborder="0" allowfullscreen></iframe>`
        } else if (url.endsWith('.mp4')) {
          newElmHtml = `<video controls autostart="0" name="media" width="640" height="360"><source src="${url}" type="video/mp4"></video>`
        } else {
          const driveMatch = url.match(rxDrive)
          if (driveMatch) {
            const driveId = driveMatch[4]
            newElmHtml = `<iframe id="driveplayer" type="text/html" width="640" height="360" src="https://drive.google.com/file/d/${driveId}/preview" frameborder="0" allowfullscreen></iframe>`
          } else {
            return
          }
        }

        const newElm = document.createElement('div') 
        newElm.classList.add('responsive-embed')
        newElm.insertAdjacentHTML('beforeend', newElmHtml)
        elm.replaceWith(newElm)
      })
    }  
  })
</script>
```

**Comment by bjoliveau** [comment]
Do you have similar advice to embed video from Google Drive ?

**Comment by tr1age** [comment]
I am unable to get this to work for some reason on my server. All youtube embeds just don't show up. I put your code not only in the HTML inject but in the seperate page as well to see. Nada. Any help would be much appreciated!

**Reply by Philipinho** [reply]
Could you please verify the Chrome console for any potential errors when you load the page and inform me about them?

**Reply by MildlyInterested** [reply]
```
<script type="text/javascript">
  function replaceGermanLocals() {
    var x = document.getElementsByClassName("text-none");
    var i;
    for (i = 0; i < x.length; i++) {
      x[i].innerText = x[i].innerText.replace(/Kommentar veröffentlichen/g, 'OK');
    }
  }

  const rxYoutube = /^.*^((?:https?:)?\/\/)?((?:www|m)\.)?((?:youtube\.com|youtu.be))(\/(?:[\w\-]+\?v=|embed\/|v\/|shorts\/)?)([\w\-]+)(\S+)?$/
  const rxScreencast = /^.*^((?:https?:)?\/\/)?(www\.)?(screencast\.com)(\/users)\/([a-z0-9_-]+)\/folders\/([a-z0-9%_-]+)\/media\/([a-z0-9_-]+)(?:\/)?$/im

  function handleEmbeddedMedia() {
    document.querySelectorAll('.contents oembed, .contents a').forEach(elm => {
      const url = elm.hasAttribute('url') ? elm.getAttribute('url') : elm.getAttribute('href');
      let newElmHtml = null;

      const ytMatch = url.match(rxYoutube);
      const scMatch = url.match(rxScreencast);

      if (ytMatch) {
        newElmHtml = `<iframe id="ytplayer" type="text/html" width="640" height="360" src="https://www.youtube-nocookie.com/embed/${ytMatch[5]}" frameborder="0" allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>`;
      } else if (scMatch) {
        newElmHtml = `<iframe id="scplayer" type="text/html" width="640" height="360" src="${url}/embed" frameborder="0" allowfullscreen></iframe>`;
      } else if (url.endsWith('.mp4')) {
        newElmHtml = `<video controls autostart="0" name="media" width="640" height="360"><source src="${url}" type="video/mp4"></video>`;
      } else {
        return;
      }

      const newElm = document.createElement('div');
      newElm.classList.add('responsive-embed');
      newElm.insertAdjacentHTML('beforeend', newElmHtml);
      elm.replaceWith(newElm);
    });
  }

  window.boot.register('page-ready', () => {
    replaceGermanLocals();
    handleEmbeddedMedia();    
  })
</script>
```
into `Head HTML  injection` found in the "Theme" section and insert your youtube link via CTRL+K.
![image](https://user-images.githubusercontent.com/57712666/227721362-5dd68347-ed65-4a9e-ae95-8be207c4e35e.png)

**Reply by Philipinho** [reply]
@MildlyInterested I hope it worked well for you. Although, you may not need the "replaceGermanLocals();".

**Comment by madman9999** [comment]
please delete

**Comment by Mubeen142** [comment]
There's a much easier way to do this

Add the following to where you want the iframe to load on the page, wiki js will give it an id of `youtube-video`

```
### YouTube Video 
```

Then click on page, click on scripts, here you can define the page script for each page

```html
<script>
window.onload = function() {
    document.getElementById('youtube-video').outerHTML = '<iframe width="100%" height="550" src="https://www.youtube.com/embed/Q-0T3uA225Y?si=golv8OHBelfG_mXq" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>';
};
</script>
```

Add this code into the script. Replace the URL with your youtube's URL

**Reply by mittkkoo** [reply]
Well this is definitely not easier. With the method described above in this thread you only need to paste the Youtube URL and that's it.
But I will note down the tweak to get an element id by defining a subcaption!

**Comment by AlexBor-89** [comment]
Версия кода для работы с RuTube:

Переопределение CSS (Без изменений):
```
.responsive-embed {
  position: relative;
  padding-bottom: 56.2%;
  height: 0;
  margin: 10px 0;
  overflow: hidden
}
.responsive-embed iframe {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border-radius: 15px
}
```

А вот внедрение в HTML:
```
<script type="text/javascript" defer>
  const rxYoutube = /^.*^((?:https?:)?\/\/)?((?:www|m)\.)?((?:youtube\.com|youtu.be))(\/(?:[\w\-]+\?v=|embed\/|v\/|shorts\/)?)([\w\-]+)(\S+)?$/
  const rxScreencast = /^.*^((?:https?:)?\/\/)?(www\.)?(screencast\.com)(\/users)\/([a-z0-9_-]+)\/folders\/([a-z0-9%_-]+)\/media\/([a-z0-9_-]+)(?:\/)?$/im
  const rxDrive = /^.*^((?:https?:)?\/\/)?((?:www|m)\.)?(drive\.google\.com)\/file\/d\/([a-zA-Z0-9_-]+)\/.*$/i
  const rxRuTube = /^.*^((?:https?:)?\/\/)?((?:www|m)\.)?(rutube\.ru)\/video\/([\w\-]+)(\S+)?$/

  window.boot.register('vue', () => {
    window.onload = () => {
      document.querySelectorAll('.contents oembed, .contents a').forEach(elm => {
        const url = elm.hasAttribute('url') ? elm.getAttribute('url') : elm.getAttribute('href')
        let newElmHtml = null
        const ytMatch = url.match(rxYoutube)
        const scMatch = url.match(rxScreencast)
        const rtMatch = url.match(rxRuTube)

        if (ytMatch) {
          newElmHtml = `<iframe id="ytplayer" type="text/html" width="640" height="360" src="https://www.youtube-nocookie.com/embed/${ytMatch[5]}" frameborder="0" allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>`
        } else if (scMatch) {
          newElmHtml = `<iframe id="scplayer" type="text/html" width="640" height="360" src="${url}/embed" frameborder="0" allowfullscreen></iframe>`
        } else if (rtMatch) {
          newElmHtml = `<iframe id="rtplayer" type="text/html" width="640" height="360" src="https://rutube.ru/play/embed/${rtMatch[4]}" frameborder="0" allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>`
        } else if (url.endsWith('.mp4')) {
          newElmHtml = `<video controls autostart="0" name="media" width="640" height="360"><source src="${url}" type="video/mp4"></video>`
        } else {
          const driveMatch = url.match(rxDrive)
          if (driveMatch) {
            const driveId = driveMatch[4]
            newElmHtml = `<iframe id="driveplayer" type="text/html" width="640" height="360" src="https://drive.google.com/file/d/${driveId}/preview" frameborder="0" allowfullscreen></iframe>`
          } else {
            return
          }
        }

        const newElm = document.createElement('div')
        newElm.classList.add('responsive-embed')
        newElm.insertAdjacentHTML('beforeend', newElmHtml)
        elm.replaceWith(newElm)
      })
    }
  })
</script>
```

**Comment by BPowell76** [comment]
If you enable `iframe` usage on your site, you can just use the `<iframe></iframe>` block in markdown to render the video. [W3C](https://www.w3schools.com/tags/tag_iframe.asp) outlines more of how to display a video using iframes.

If using YouTube as your video source, they generate the iframe tag for you if you go to **Share** > **Embed**

**Comment by Maelstromeous** [comment]
I can confirm if you do this, you can use embeds for YouTube:

1. Open Admin Settings
2. Go to Modules -> Rendering
3. Go to HTML
4. Security
5. Enable Allow iFrames.

In order to use this, you need to convert your page to raw HTML. For most people this may a bridge too far, but it's pretty simple to learn.

When the page has been converted to HTML, you can then go to the YouTube video you wish to embed, press Share, then Embed and there's some code for you to copy and paste in there.

Save the page, and it should render. If the video says "Watch on YouTube" they have turned off embedding (usually cos adverts).
